File: services/api/src/pipeline.js
/**
 * The two long-running operations: ingest and render.
 *
 * Kept out of the HTTP layer so they can be tested directly and, later, moved
 * behind a queue without touching the routes.
 */

import fs from 'node:fs/promises';
import path from 'node:path';
import { buildEdl, diffTranscript, identityTokens } from '@voxdocs/edl';
import { makePreviewAudio, normalizeToMaster, probe } from './ffmpeg.js';
import * as model from './modelClient.js';
import { encodeAudio, expandSegments, opDurations, renderOps, renderVideo } from './render.js';
import { newId, projectPaths, save } from './store.js';

const AUDIO_FORMATS = new Set(['wav', 'mp3', 'm4a']);

export class ValidationError extends Error {
  /** @param {string} message */
  constructor(message) {
    super(message);
    this.name = 'ValidationError';
    this.status = 400;
  }
}

/**
 * Decode, measure, transcribe. Ingest is the only place the media is inspected;
 * everything downstream works from the master and the transcript.
 *
 * @param {any} project
 * @param {string} sourcePath
 * @param {{ language?: string | null }} [options]
 */
export async function ingest(project, sourcePath, options = {}) {
  const paths = projectPaths(project.id);

  const info = await probe(sourcePath);
  if (!info.hasAudio) {
    throw new ValidationError('this file has no audio track to transcribe');
  }

  project.duration = info.duration;
  project.hasVideo = info.hasVideo;
  project.video = info.hasVideo
    ? { width: info.width, height: info.height, fps: Number(info.fps.toFixed(3)) }
    : null;
  project.status = 'transcribing';
  await save(project);

  // One canonical master drives every later cut.
  await normalizeToMaster(sourcePath, paths.master);
  try {
    await makePreviewAudio(sourcePath, paths.preview);
  } catch {
    // A missing preview only costs the editor its scrub audio; not fatal.
    project.previewFailed = true;
  }

  const result = await model.transcribe(paths.master, project.id, options.language ?? null);

  project.transcript = {
    words: result.words ?? [],
    segments: result.segments ?? [],
    language: result.language ?? 'en',
    backend: result.backend ?? 'unknown',
  };
  project.voice = result.voice ?? null;
  project.duration = result.duration || project.duration;
  project.status = 'ready';
  project.error = null;

  // The envelope is large and only the editor wants it; keep it out of the
  // project document so every other read stays cheap.
  await fs.writeFile(
    path.join(paths.dir, 'envelope.json'),
    JSON.stringify(result.envelope ?? { fps: 100, rms: [] }),
    'utf8'
  );

  await save(project);
  return project;
}

/** @param {any} project */
export async function loadEnvelope(project) {
  const paths = projectPaths(project.id);
  try {
    return JSON.parse(await fs.readFile(path.join(paths.dir, 'envelope.json'), 'utf8'));
  } catch {
    return null;
  }
}

/**
 * Turn a request body into an edit-token list.
 *
 * Two shapes are accepted: the structured list the editor maintains (exact, no
 * guessing), and plain text (aligned against the original). The editor uses the
 * first; pastes, scripts and API clients use the second.
 *
 * @param {any} project
 * @param {any} body
 */
export function resolveTokens(project, body) {
  const words = project.transcript?.words ?? [];

  if (Array.isArray(body?.tokens)) {
    const tokens = [];
    for (const token of body.tokens) {
      if (token && typeof token.ref === 'string') tokens.push({ ref: token.ref });
      else if (token && typeof token.insert === 'string') tokens.push({ insert: token.insert });
      else throw new ValidationError('each token must be {ref:string} or {insert:string}');
    }
    return tokens;
  }

  if (typeof body?.text === 'string') {
    return diffTranscript(words, body.text);
  }

  return identityTokens(words);
}

/**
 * Build the EDL without rendering, for live duration and cost feedback.
 * @param {any} project
 * @param {Array<{ ref: string } | { insert: string }>} tokens
 */
export async function planEdit(project, tokens) {
  const envelope = await loadEnvelope(project);
  return buildEdl(project.transcript?.words ?? [], tokens, {
    duration: project.duration,
    envelope,
  });
}

/**
 * Render an edit to a finished file.
 *
 * @param {any} project
 * @param {Array<{ ref: string } | { insert: string }>} tokens
 * @param {{ format?: string, video?: boolean }} [options]
 */
export async function render(project, tokens, options = {}) {
  const format = String(options.format ?? 'wav').toLowerCase();
  const wantVideo = Boolean(options.video) && Boolean(project.hasVideo);
  if (!wantVideo && !AUDIO_FORMATS.has(format)) {
    throw new ValidationError(`unsupported format "${format}" (use wav, mp3 or m4a)`);
  }

  const paths = projectPaths(project.id);
  const { segments, stats } = await planEdit(project, tokens);

  // Ask the model server for every insertion at once.
  const synthIndices = [];
  const items = [];
  segments.forEach((segment, index) => {
    if (segment.kind !== 'synth') return;
    synthIndices.push(index);
    items.push({
      text: segment.text,
      context_before: segment.contextBefore,
      context_after: segment.contextAfter,
      lead_gap: segment.leadGap,
      trail_gap: segment.trailGap,
    });
  });

  const synthesis = new Map();
  if (items.length > 0) {
    const reseed = async () => {
      await model.putVoiceProfile(
        project.id,
        project.transcript.words,
        project.duration,
        paths.master
      );
    };
    const response = await model.synthesizeBatch(project.id, items, reseed);
    (response.results ?? []).forEach((result, i) => synthesis.set(synthIndices[i], result));
  }

  await fs.mkdir(paths.renders, { recursive: true });
  const renderId = newId();
  const workDir = path.join(paths.renders, renderId);
  await fs.mkdir(workDir, { recursive: true });

  try {
    const { ops, warnings } = await expandSegments(segments, synthesis, workDir);

    const masterOut = path.join(workDir, 'render.wav');
    const rendered = await renderOps(paths.master, ops, masterOut);

    // Video needs to know how long each insertion actually turned out to be.
    const synthDurations = new Map();
    if (wantVideo) {
      const durations = await opDurations(ops);
      let cursor = 0;
      for (let i = 0; i < segments.length; i++) {
        if (segments[i].kind === 'copy') {
          cursor += 1;
          continue;
        }
        const result = synthesis.get(i);
        const count = (result?.units ?? []).filter((u) => u.type !== 'silence' || u.duration > 0.001)
          .length;
        let total = 0;
        for (let k = 0; k < count && cursor < durations.length; k++, cursor++) {
          total += durations[cursor];
        }
        synthDurations.set(i, total);
      }
    }

    let outputPath;
    let outputFormat;
    if (wantVideo) {
      outputPath = path.join(workDir, 'output.mp4');
      outputFormat = 'mp4';
      const sourcePath = await findSource(project);
      await renderVideo(sourcePath, masterOut, segments, synthDurations, outputPath);
    } else {
      outputFormat = format;
      outputPath = path.join(workDir, `output.${format}`);
      await encodeAudio(masterOut, outputPath, /** @type {any} */ (format));
    }

    // The intermediate master and any inline TTS clips have served their purpose.
    await fs.rm(masterOut, { force: true });
    for (const entry of await fs.readdir(workDir)) {
      if (entry.startsWith('tts-')) await fs.rm(path.join(workDir, entry), { force: true });
    }

    const { size } = await fs.stat(outputPath);
    const record = {
      id: renderId,
      createdAt: new Date().toISOString(),
      format: outputFormat,
      file: path.basename(outputPath),
      bytes: size,
      duration: rendered.duration,
      pieces: rendered.pieces,
      stats,
      warnings,
      synthesis: summariseSynthesis(synthesis),
    };

    project.renders = [record, ...(project.renders ?? [])].slice(0, 25);
    await save(project);
    return record;
  } catch (error) {
    await fs.rm(workDir, { recursive: true, force: true });
    throw error;
  }
}

/** @param {Map<number, any>} synthesis */
function summariseSynthesis(synthesis) {
  const covered = [];
  const generated = [];
  const missing = [];
  const backends = new Set();
  for (const result of synthesis.values()) {
    covered.push(...(result.covered ?? []));
    generated.push(...(result.generated ?? []));
    missing.push(...(result.missing ?? []));
    for (const backend of result.backends ?? []) backends.add(backend);
  }
  const total = covered.length + generated.length + missing.length;
  return {
    words: total,
    fromVoiceBank: covered.length,
    fromTts: generated.length,
    missing,
    backends: [...backends],
    coverage: total ? Number((covered.length / total).toFixed(4)) : 1,
  };
}

/** Locate the original upload, whose extension varies. */
async function findSource(project) {
  const paths = projectPaths(project.id);
  if (project.sourceFile) {
    const candidate = path.join(paths.dir, project.sourceFile);
    try {
      await fs.access(candidate);
      return candidate;
    } catch {
      /* fall through to a scan */
    }
  }
  for (const entry of await fs.readdir(paths.dir)) {
    if (entry.startsWith('source.')) return path.join(paths.dir, entry);
  }
  throw new ValidationError('the original media file is no longer available');
}

File: services/backend/projects/services/pipeline.py
"""The two long-running operations: ingest and render.

Kept out of both the HTTP layer and the Celery layer so they can be called
directly from a test or a management command, and so the task functions stay
thin enough to be obviously correct.
"""

from __future__ import annotations

import json
import logging
import shutil
from pathlib import Path

from django.conf import settings
from django.db import transaction

from edl import CopySegment, SynthSegment, build_edl, diff_transcript, identity_tokens

from ..models import Project, Render, Segment, Word
from . import ffmpeg, model_client
from .render import encode_audio, expand_segments, op_durations, render_ops, render_video

log = logging.getLogger(__name__)

AUDIO_FORMATS = {"wav", "mp3", "m4a"}


class ValidationError(Exception):
    """The request cannot be satisfied as asked. Surfaces as a 400."""

    status = 400


# ------------------------------------------------------------------ ingest

def ingest(project: Project, language: str | None = None) -> Project:
    """Decode, measure, transcribe.

    The only place the media is inspected; everything downstream works from the
    master and the stored transcript.
    """
    source = project.source_path()
    if source is None:
        raise ValidationError("the uploaded file is missing")

    info = ffmpeg.probe(source)
    if not info["has_audio"]:
        raise ValidationError("this file has no audio track to transcribe")

    project.duration = info["duration"]
    project.has_video = info["has_video"]
    project.video_width = info["width"]
    project.video_height = info["height"]
    project.video_fps = round(info["fps"], 3)
    project.status = Project.Status.TRANSCRIBING
    project.save(update_fields=[
        "duration", "has_video", "video_width", "video_height", "video_fps",
        "status", "updated_at",
    ])

    project.directory.mkdir(parents=True, exist_ok=True)

    # One canonical master drives every later cut.
    ffmpeg.normalize_to_master(source, project.master_path)
    try:
        ffmpeg.make_preview_audio(source, project.preview_path)
        project.preview_failed = False
    except ffmpeg.FfmpegError:
        # A missing preview only costs the editor its scrub audio; not fatal.
        log.warning("preview encode failed for %s", project.id)
        project.preview_failed = True

    result = model_client.transcribe(project.master_path, project.id, language)

    words = result.get("words") or []
    segments = result.get("segments") or []
    segment_of = {}
    for index, segment in enumerate(segments):
        for i in range(segment.get("first_word", 0), segment.get("last_word", -1) + 1):
            segment_of[i] = index

    with transaction.atomic():
        project.words.all().delete()
        project.segments.all().delete()
        Word.objects.bulk_create([
            Word(
                project=project,
                index=i,
                token_id=str(word.get("id") or f"w{i}"),
                text=str(word.get("text", ""))[:255],
                start=float(word.get("start", 0.0)),
                end=float(word.get("end", 0.0)),
                confidence=float(word.get("confidence", 1.0)),
                segment=segment_of.get(i, 0),
            )
            for i, word in enumerate(words)
        ], batch_size=1000)
        Segment.objects.bulk_create([
            Segment(
                project=project,
                index=index,
                start=float(segment.get("start", 0.0)),
                end=float(segment.get("end", 0.0)),
                text=segment.get("text", ""),
                first_word=int(segment.get("first_word", 0)),
                last_word=int(segment.get("last_word", 0)),
            )
            for index, segment in enumerate(segments)
        ], batch_size=500)

        project.language = result.get("language") or "en"
        project.asr_backend = result.get("backend") or "unknown"
        project.voice = result.get("voice")
        project.duration = result.get("duration") or project.duration
        project.status = Project.Status.READY
        project.error = ""
        project.save()

    # The envelope is large and only the editor wants it, so it stays a file
    # rather than bloating every read of the project row.
    project.envelope_path.write_text(
        json.dumps(result.get("envelope") or {"fps": 100, "rms": []}), encoding="utf-8"
    )
    return project


def load_envelope(project: Project) -> dict | None:
    try:
        return json.loads(project.envelope_path.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return None


# ------------------------------------------------------------------- edits

def resolve_tokens(project: Project, payload: dict) -> list[dict]:
    """Turn a request body into an edit-token list.

    Two shapes are accepted: the structured list the editor maintains (exact, no
    guessing), and plain text (aligned against the original). The editor uses the
    first; pastes, scripts and API clients use the second.
    """
    raw_tokens = payload.get("tokens")
    if isinstance(raw_tokens, list):
        tokens = []
        for token in raw_tokens:
            if isinstance(token, dict) and isinstance(token.get("ref"), str):
                tokens.append({"ref": token["ref"]})
            elif isinstance(token, dict) and isinstance(token.get("insert"), str):
                tokens.append({"insert": token["insert"]})
            else:
                raise ValidationError("each token must be {ref:string} or {insert:string}")
        return tokens

    text = payload.get("text")
    if isinstance(text, str):
        return diff_transcript(project.edl_words(), text)

    return identity_tokens(project.edl_words())


def plan_edit(project: Project, tokens: list[dict]):
    """Build the EDL without rendering, for live duration feedback."""
    return build_edl(
        project.edl_words(), tokens,
        duration=project.duration,
        envelope=load_envelope(project),
    )


# ------------------------------------------------------------------ render

def perform_render(render: Render) -> Render:
    """Execute a queued render. Called from the Celery task."""
    project = render.project
    fmt = (render.format or "wav").lower()
    want_video = bool(render.with_video and project.has_video)
    if not want_video and fmt not in AUDIO_FORMATS:
        raise ValidationError(f'unsupported format "{fmt}" (use wav, mp3 or m4a)')

    edl = plan_edit(project, render.tokens)

    # Ask the model server for every insertion at once.
    synth_indices = []
    items = []
    for index, segment in enumerate(edl.segments):
        if not isinstance(segment, SynthSegment):
            continue
        synth_indices.append(index)
        items.append({
            "text": segment.text,
            "context_before": segment.context_before,
            "context_after": segment.context_after,
            "lead_gap": segment.lead_gap,
            "trail_gap": segment.trail_gap,
            "quality": render.quality,
        })

    synthesis: dict[int, dict] = {}
    if items:
        def reseed():
            model_client.put_voice_profile(
                project.id,
                [
                    {"text": w["text"], "start": w["start"], "end": w["end"],
                     "confidence": w.get("confidence", 1.0)}
                    for w in project.words.values("text", "start", "end", "confidence")
                ],
                project.duration,
                project.master_path,
            )

        response = model_client.synthesize_batch(project.id, items, reseed)
        for i, result in enumerate(response.get("results", [])):
            synthesis[synth_indices[i]] = result

    workdir = render.directory
    workdir.mkdir(parents=True, exist_ok=True)

    try:
        ops, warnings = expand_segments(edl.segments, synthesis, workdir)

        master_out = workdir / "render.wav"
        rendered = render_ops(project.master_path, ops, master_out)

        # Video needs to know how long each insertion actually turned out.
        synth_durations: dict[int, float] = {}
        if want_video:
            durations = op_durations(ops)
            cursor = 0
            for index, segment in enumerate(edl.segments):
                if isinstance(segment, CopySegment):
                    cursor += 1
                    continue
                units = synthesis.get(index, {}).get("units", [])
                count = sum(
                    1 for u in units
                    if u.get("type") != "silence" or float(u.get("duration", 0)) > 0.001
                )
                total = 0.0
                for _ in range(count):
                    if cursor >= len(durations):
                        break
                    total += durations[cursor]
                    cursor += 1
                synth_durations[index] = total

        if want_video:
            output = workdir / "output.mp4"
            output_format = "mp4"
            source = project.source_path()
            if source is None:
                raise ValidationError("the original media file is no longer available")
            render_video(source, master_out, edl.segments, synth_durations, output)
        else:
            output_format = fmt
            output = workdir / f"output.{fmt}"
            encode_audio(master_out, output, fmt)

        # The intermediate master and inline TTS clips have served their purpose.
        master_out.unlink(missing_ok=True)
        for entry in workdir.glob("tts-*"):
            entry.unlink(missing_ok=True)

        render.status = Render.Status.READY
        render.format = output_format
        render.file = output.name
        render.bytes = output.stat().st_size
        render.duration = rendered["duration"]
        render.pieces = rendered["pieces"]
        render.stats = edl.stats.to_json()
        render.warnings = warnings
        render.synthesis = summarise_synthesis(synthesis)
        render.error = ""
        render.save()
        return render
    except Exception:
        shutil.rmtree(workdir, ignore_errors=True)
        raise


def summarise_synthesis(synthesis: dict[int, dict]) -> dict:
    covered, generated, missing = [], [], []
    backends: list[str] = []
    for result in synthesis.values():
        covered += result.get("covered", [])
        generated += result.get("generated", [])
        missing += result.get("missing", [])
        for backend in result.get("backends", []):
            if backend not in backends:
                backends.append(backend)

    total = len(covered) + len(generated) + len(missing)
    return {
        "words": total,
        "fromVoiceBank": len(covered),
        "fromTts": len(generated),
        "missing": missing,
        "backends": backends,
        "coverage": round(len(covered) / total, 4) if total else 1,
    }


def delete_project_files(project: Project) -> None:
    shutil.rmtree(project.directory, ignore_errors=True)
